File: examples/anthropic-managed-agents/python/anthropic_managed_agents_e2b/environment.py
from __future__ import annotations

import json

import anthropic

CONSOLE_URL = "https://platform.claude.com/workspaces/default/environments/{environment_id}"
WORKER_SANDBOX_METADATA_KEY = "e2b_worker_sandbox_id"
WEBHOOK_SANDBOX_METADATA_KEY = "e2b_webhook_sandbox_id"
WORKER_SANDBOX_STORE_METADATA_KEY = "e2b_worker_sandbox_ids"
WEBHOOK_SANDBOX_STORE_METADATA_KEY = "e2b_webhook_sandbox_ids"


def create_self_hosted_environment(*, api_key: str, name: str):
    client = anthropic.Anthropic(api_key=api_key)
    return client.beta.environments.create(
        name=name,
        config={"type": "self_hosted"},
    )


def retrieve_environment(*, api_key: str, environment_id: str):
    client = anthropic.Anthropic(api_key=api_key)
    return client.beta.environments.retrieve(environment_id)


def update_environment_metadata(
    *,
    api_key: str,
    environment_id: str,
    metadata: dict[str, str | None],
):
    client = anthropic.Anthropic(api_key=api_key)
    return client.beta.environments.update(
        environment_id,
        metadata=metadata,
    )


def sandbox_store(metadata: dict[str, str], *, store_key: str, legacy_key: str) -> list[str]:
    ids: list[str] = []
    raw = metadata.get(store_key)
    if raw:
        try:
            parsed = json.loads(raw)
        except json.JSONDecodeError:
            parsed = []
        if isinstance(parsed, list):
            ids.extend(item for item in parsed if isinstance(item, str) and item)

    legacy_id = metadata.get(legacy_key)
    if legacy_id:
        ids.append(legacy_id)

    return list(dict.fromkeys(ids))


def serialize_sandbox_store(sandbox_ids: list[str]) -> str:
    return json.dumps(list(dict.fromkeys(sandbox_ids)), separators=(",", ":"))


def add_sandbox_to_metadata_store(
    *,
    api_key: str,
    environment_id: str,
    store_key: str,
    legacy_key: str,
    sandbox_id: str,
):
    env = retrieve_environment(api_key=api_key, environment_id=environment_id)
    sandbox_ids = sandbox_store(env.metadata, store_key=store_key, legacy_key=legacy_key)
    sandbox_ids = [sandbox_id, *(item for item in sandbox_ids if item != sandbox_id)]
    return update_environment_metadata(
        api_key=api_key,
        environment_id=environment_id,
        metadata={
            legacy_key: sandbox_id,
            store_key: serialize_sandbox_store(sandbox_ids),
        },
    )


def clear_matching_sandbox_metadata(
    *,
    api_key: str,
    environment_id: str,
    sandbox_id: str,
):
    env = retrieve_environment(api_key=api_key, environment_id=environment_id)
    metadata: dict[str, str | None] = {}
    for legacy_key, store_key in (
        (WORKER_SANDBOX_METADATA_KEY, WORKER_SANDBOX_STORE_METADATA_KEY),
        (WEBHOOK_SANDBOX_METADATA_KEY, WEBHOOK_SANDBOX_STORE_METADATA_KEY),
    ):
        if env.metadata.get(legacy_key) == sandbox_id:
            metadata[legacy_key] = None
        sandbox_ids = sandbox_store(env.metadata, store_key=store_key, legacy_key=legacy_key)
        updated_ids = [item for item in sandbox_ids if item != sandbox_id]
        if updated_ids != sandbox_ids:
            metadata[store_key] = serialize_sandbox_store(updated_ids)
    if not metadata:
        return env
    return update_environment_metadata(
        api_key=api_key,
        environment_id=environment_id,
        metadata=metadata,
    )


def console_url(environment_id: str) -> str:
    return CONSOLE_URL.format(environment_id=environment_id)

File: examples/anthropic-managed-agents/javascript/src/environment.ts
import Anthropic from "@anthropic-ai/sdk";

const CONSOLE_URL = "https://platform.claude.com/workspaces/default/environments";
export const WORKER_SANDBOX_METADATA_KEY = "e2b_worker_sandbox_id";
export const WEBHOOK_SANDBOX_METADATA_KEY = "e2b_webhook_sandbox_id";
export const WORKER_SANDBOX_STORE_METADATA_KEY = "e2b_worker_sandbox_ids";
export const WEBHOOK_SANDBOX_STORE_METADATA_KEY = "e2b_webhook_sandbox_ids";

export async function createSelfHostedEnvironment({
  apiKey,
  name,
}: {
  apiKey: string;
  name: string;
}) {
  const client = new Anthropic({ apiKey });
  return client.beta.environments.create({
    name,
    config: { type: "self_hosted" },
  });
}

export async function retrieveEnvironment({
  apiKey,
  environmentId,
}: {
  apiKey: string;
  environmentId: string;
}) {
  const client = new Anthropic({ apiKey });
  return client.beta.environments.retrieve(environmentId);
}

export async function updateEnvironmentMetadata({
  apiKey,
  environmentId,
  metadata,
}: {
  apiKey: string;
  environmentId: string;
  metadata: Record<string, string | null>;
}) {
  const client = new Anthropic({ apiKey });
  return client.beta.environments.update(environmentId, { metadata });
}

export function sandboxStore({
  metadata,
  storeKey,
  legacyKey,
}: {
  metadata: Record<string, string>;
  storeKey: string;
  legacyKey: string;
}) {
  const sandboxIds: string[] = [];
  const raw = metadata[storeKey];
  if (raw) {
    try {
      const parsed = JSON.parse(raw);
      if (Array.isArray(parsed)) {
        sandboxIds.push(...parsed.filter((item): item is string => typeof item === "string" && Boolean(item)));
      }
    } catch {
      // Ignore malformed metadata and fall back to the legacy key.
    }
  }

  const legacyId = metadata[legacyKey];
  if (legacyId) {
    sandboxIds.push(legacyId);
  }

  return [...new Set(sandboxIds)];
}

function serializeSandboxStore(sandboxIds: string[]) {
  return JSON.stringify([...new Set(sandboxIds)]);
}

export async function addSandboxToMetadataStore({
  apiKey,
  environmentId,
  storeKey,
  legacyKey,
  sandboxId,
}: {
  apiKey: string;
  environmentId: string;
  storeKey: string;
  legacyKey: string;
  sandboxId: string;
}) {
  const environment = await retrieveEnvironment({ apiKey, environmentId });
  const sandboxIds = sandboxStore({ metadata: environment.metadata, storeKey, legacyKey });
  return updateEnvironmentMetadata({
    apiKey,
    environmentId,
    metadata: {
      [legacyKey]: sandboxId,
      [storeKey]: serializeSandboxStore([
        sandboxId,
        ...sandboxIds.filter((item) => item !== sandboxId),
      ]),
    },
  });
}

export async function clearMatchingSandboxMetadata({
  apiKey,
  environmentId,
  sandboxId,
}: {
  apiKey: string;
  environmentId: string;
  sandboxId: string;
}) {
  const environment = await retrieveEnvironment({ apiKey, environmentId });
  const metadata: Record<string, string | null> = {};

  for (const [legacyKey, storeKey] of [
    [WORKER_SANDBOX_METADATA_KEY, WORKER_SANDBOX_STORE_METADATA_KEY],
    [WEBHOOK_SANDBOX_METADATA_KEY, WEBHOOK_SANDBOX_STORE_METADATA_KEY],
  ] as const) {
    if (environment.metadata[legacyKey] === sandboxId) {
      metadata[legacyKey] = null;
    }
    const sandboxIds = sandboxStore({ metadata: environment.metadata, storeKey, legacyKey });
    const updatedIds = sandboxIds.filter((item) => item !== sandboxId);
    if (updatedIds.length !== sandboxIds.length) {
      metadata[storeKey] = serializeSandboxStore(updatedIds);
    }
  }

  if (Object.keys(metadata).length === 0) {
    return environment;
  }

  return updateEnvironmentMetadata({ apiKey, environmentId, metadata });
}

export function consoleUrl(environmentId: string) {
  return `${CONSOLE_URL}/${environmentId}`;
}
